### ISSUE TITLE
SpazzyRuler ID

### ISSUE THREAD

**Issue by kaykhancheckpoint** [gandersen101/spaczz#17]
is it possible to add an "id" to the pattern like you can with original spacy api? https://spacy.io/usage/rule-based-matching#entityruler-ent-ids

https://github.com/gandersen101/spaczz#spaczzruler

```
 p = {
            "label": entity_type,
            "pattern": d[col],
            "id": d["id"],
            "type": "fuzzy"
        }
```

**Pattern**: {'label': 'PERSON', 'pattern': 'DrDisrespect', 'id': '148265b0-847b-414c-9f8e-916561412c55', 'type': 'fuzzy'}
**Text**: "Dr Disrespect to Returns Aug. 7 With YouTube Stream, Will Explore Other Platform Options"

This fails to find Dr Disrespect and im unable to print the id
```

    data = [{
        "label": ent.label_,
        "name": ent.text,
        "id": ent.ent_id_
    } for ent in doc.ents]

print(data)
```

I need this feature because when i find the entity i need to match it to an "id" that stored in my database. 

aside: is it also possible to edit the fuzzy search edit distance to 1 character

**Comment by gandersen101** [comment]
Hi @kaykhancheckpoint,

Thanks for your interest in spaczz. It's nice to see interested users, but I'm sorry you're having this issue.

Adding an "ent_id" like in spacy's entity ruler should be doable, just give me a couple days to implement this feature as it is only me currently working on this project. Hopefully I can have it addressed by the end of the weekend.

However, I am a little confused as to why you are unable to get a match between "Dr Disrespect" and "DrDisrespect".  The below code finds a match in both the current release and latest dev version of spaczz, although this does not incorporate the ent_id options, which is not currently implemented:

```python
nlp = spacy.blank("en")
text = "Dr Disrespect to Returns Aug. 7 With YouTube Stream, Will Explore Other Platform Options"
doc = nlp(text)

patterns = [
    {"label": "PERSON", "pattern": "DrDisrespect", "type": "fuzzy"}
]

ruler = SpaczzRuler(nlp)
ruler.add_patterns(patterns)
doc = ruler(doc)

data = [{
    "label": ent.label_,
    "name": ent.text,
} for ent in doc.ents]

print(data)
```

    [{'label': 'PERSON', 'name': 'Dr Disrespect'}]

Can you see if this works for you like above, without the "ent_id" logic?

As for you ask for search distance of 1 character, spaczz currently only searchers tokens rather than characters, although a character searching version of the matcher is doable, I would likely treat that as a separate issue.

Spaczz's fuzzy matching currently iterates through a doc one token at a time and by default will flex the boundaries of any potential matches (see the `min_r1` parameters in the `FuzzySearcher.compare()` method) by the token length of the pattern. This can be limited to only one token by change the `flex` value to 1.  You can adjust that in the optional `kwargs` of a fuzzy pattern for use in the spaczz matcher and ruler.

Let me know if you have any additional questions or concerns and I will start implementing the "ent_id" logic.

Thanks!

**Comment by kaykhancheckpoint** [comment]
Hey i managed to get that working, but I'm not sure why it wasn't working with my code base, ill take a look at it.

Anyway thanks for taking a look into the "ent_id", it's going to be be super useful for me.

**Comment by gandersen101** [comment]
Hi @kaykhancheckpoint the "ent_id" feature is now implemented in the branch, [enhancement-ent-ids](https://github.com/gandersen101/spaczz/tree/enhancement-ent-ids). I have some more housekeeping to do before merging it into master and creating a new build but I intend to have that done sometime later tomorrow.

As there will be a few other changes included in the next release (v0.2.0), please make sure to check the release notes that will be towards the top of the updated readme after the new build is released.
